pkitesting/src/main/java/io/netty/pkitesting/CertificateBuilder.java

```diff
@@ -0,0 +1,1088 @@
+
+    /**
+     * Create a new certificate builder with a default configuration.
+     * Unless specified otherwise, the builder will produce bundles that use the
+     * {@linkplain Algorithm#ecp256 NIST EC-P 256} key algorithm,
+     * and the certificates will be valid as of yesterday and expire tomorrow.
+     */
+    public CertificateBuilder() {
+    }
+
+    /**
+     * Produce a copy of the current state in this certificate builder.
+     * @return A copy of this certificate builder.
+     */
+    public CertificateBuilder copy() {
+        CertificateBuilder copy = new CertificateBuilder();
+        copy.random = random;
+        copy.algorithm = algorithm;
+        copy.notBefore = notBefore;
+        copy.notAfter = notAfter;
+        copy.modifierCallbacks = new ArrayList<>(modifierCallbacks);
+        copy.subjectAlternativeNames = new ArrayList<>(subjectAlternativeNames);
+        copy.crlDistributionPoints = new ArrayList<>(crlDistributionPoints);
+        copy.serial = serial;
+        copy.subject = subject;
+        copy.isCertificateAuthority = isCertificateAuthority;
+        copy.pathLengthConstraint = pathLengthConstraint;
+        copy.publicKey = publicKey;
+        copy.keyUsage = keyUsage;
+        copy.extendedKeyUsage = new TreeSet<>(extendedKeyUsage);
+        return copy;
+    }
+
+    /**
+     * Set the {@link SecureRandom} instance to use when generating keys.
+     * @param secureRandom The secure random instance to use.
+     * @return This certificate builder.
+     */
+    public CertificateBuilder secureRandom(SecureRandom secureRandom) {
+        random = requireNonNull(secureRandom);
+        return this;
+    }
+
+    /**
+     * Set the not-before field of the certificate. The certificate will not be valid before this time.
+     * @param instant The not-before time.
+     * @return This certificate builder.
+     */
+    public CertificateBuilder notBefore(Instant instant) {
+        notBefore = requireNonNull(instant);
+        return this;
+    }
+
+    /**
+     * Set the not-after field of the certificate. The certificate will not be valid after this time.
+     * @param instant The not-after time.
+     * @return This certificate builder.
+     */
+    public CertificateBuilder notAfter(Instant instant) {
+        notAfter = requireNonNull(instant);
+        return this;
+    }
+
+    /**
+     * Set the specific serial number to use in the certificate.
+     * One will be generated randomly, if none is specified.
+     * @param serial The serial number to use, or {@code null}.
+     * @return This certificate builder.
+     */
+    public CertificateBuilder serial(BigInteger serial) {
+        this.serial = serial;
+        return this;
+    }
+
+    /**
+     * Set the fully-qualified domain name (an X.500 name) as the subject of the certificate.
+     * @param fqdn The subject name to use.
+     * @return This certificate builder.
+     */
+    public CertificateBuilder subject(String fqdn) {
```

Comment:
How about we pass `enum` for `cn` instead of parsing it from `cn=localhost`?

Reply:
The FQDN is a directory name and can have many elements. I think `String` and `X500Principal` are the best types for capturing this. Directory names are too complicated to be represented by an enum, or enum + string, or even a sequence of enum/string pairs.